File: aoc2025/python/day3.py
import time
from utilities import read_file


class BatteryPack:
    def __init__(self, batteries):
        self.batteries = batteries

    def get_largest_pack_joltage(self, battery_size):
        max_joltage = []

        index = 0
        iteration = 0
        while True:
            if len(max_joltage) == battery_size:
                break
            max_found_joltage, end_index = self.iterate_battery_pack_check(
                index, iteration, battery_size
            )
            max_joltage.append(max_found_joltage)
            index = end_index + 1
            iteration += 1

        result = int(''.join(str(j) for j in max_joltage))
        return result

    def iterate_battery_pack_check(self, starting_index, current_battery_size, battery_size):
        max_found_joltage = 0
        end_index = starting_index
        stop = len(self.batteries) + current_battery_size - (battery_size - 1)

        for i in range(starting_index, stop):
            if self.batteries[i] > max_found_joltage:
                max_found_joltage = self.batteries[i]
                end_index = i

        return max_found_joltage, end_index


def process_day3_file():
    lines = read_file("day3.part1")
    battery_packs = []

    for line in lines:
        batteries = [int(char) for char in line]
        battery_packs.append(BatteryPack(batteries))

    return battery_packs


def calculate_12_pack_joltage(battery_packs, battery_size):
    max_joltage = 0
    for battery_pack in battery_packs:
        max_joltage += battery_pack.get_largest_pack_joltage(battery_size)

    return max_joltage


def day3():
    battery_packs = process_day3_file()

    start = time.time()
    part1_result = calculate_12_pack_joltage(battery_packs, 2)
    part1_time = time.time() - start

    start = time.time()
    part2_result = calculate_12_pack_joltage(battery_packs, 12)
    part2_time = time.time() - start

    print(f"Day 3:")
    print(f"  Part 1: {part1_result} ({part1_time:.4f}s)")
    print(f"  Part 2: {part2_result} ({part2_time:.4f}s)")


if __name__ == "__main__":
    day3()

File: aoc2025/javascript/day3.js
import { readFile } from './utilities.js';

function processDay3File() {
  const lines = readFile('day3.part1');
  const batteryPacks = [];

  for (const line of lines) {
    const batteryPack = [];
    for (const batteryChar of line) {
      const battery = parseInt(batteryChar);
      batteryPack.push(battery);
    }
    batteryPacks.push(batteryPack);
  }

  return batteryPacks;
}

function iterateBatteryPackCheck(battery, startingIndex, currentBatterySize, batterySize) {
  let maxFoundJoltage = 0;
  let endIndex = startingIndex;
  const stop = battery.length + currentBatterySize - (batterySize - 1);

  for (let i = startingIndex; i < stop; i++) {
    if (battery[i] > maxFoundJoltage) {
      maxFoundJoltage = battery[i];
      endIndex = i;
    }
  }

  return [maxFoundJoltage, endIndex];
}

function getLargestPackJoltage(batteryPack, batterySize) {
  const maxJoltage = [];

  let index = 0;
  let iteration = 0;

  while (maxJoltage.length < batterySize) {
    const [maxFoundJoltage, endIndex] = iterateBatteryPackCheck(
      batteryPack,
      index,
      iteration,
      batterySize
    );
    maxJoltage.push(maxFoundJoltage);
    index = endIndex + 1;
    iteration++;
  }

  return parseInt(maxJoltage.join(''));
}

function calculate12PackJoltage(batteryPacks, batterySize) {
  let maxJoltage = 0;
  for (const batteryPack of batteryPacks) {
    maxJoltage += getLargestPackJoltage(batteryPack, batterySize);
  }
  return maxJoltage;
}

function day3() {
  const batteryPacks = processDay3File();

  let start = performance.now();
  const part1Result = calculate12PackJoltage(batteryPacks, 2);
  const part1Time = ((performance.now() - start) / 1000).toFixed(4);

  start = performance.now();
  const part2Result = calculate12PackJoltage(batteryPacks, 12);
  const part2Time = ((performance.now() - start) / 1000).toFixed(4);

  console.log(`\nDay 3:`);
  console.log(`  Part 1: ${part1Result} (${part1Time}s)`);
  console.log(`  Part 2: ${part2Result} (${part2Time}s)`);
}

day3();
